POS API smartGrid → action Codemod

ShadiBitaraf opened the pull request:
### Overview
Adds a codemod to migrate Shopify POS extensions from the deprecated  
`api.smartGrid.presentModal()` → `api.action.presentModal()`.

### Key Features
- Transforms only `presentModal` calls, preserving other smartGrid methods
- Handles aliases, destructuring, optional chaining, and class-based usage
- AST-based, syntax-aware transformations with validation
- Backed by new utils library (`utils/ast-utils.ts`)

### Impact
- Automates migration of POS extensions with safe, tested transformations
- Reduces manual effort and risk of migration errors
- Provides reusable utilities for future codemods

### Files Added
- `utils/ast-utils.ts`  
- `codemods/pos-api-smartgrid-to-action/` (implementation, tests, docs)

ShadiBitaraf committed:
feat: initialize POS API smartGrid migration codemod project

- Set up basic codemod structure with workflow and package configuration
- Add TypeScript configuration for development
- Implement AST-based smartGrid to action transformation
- Add basic test fixtures for validation

ShadiBitaraf committed:
feat(utils): add comprehensive AST utility library

- Add getImports() with support for all import patterns
- Add getNamedImports() for specific import extraction
- Add getVariableAliases() for dynamic alias detection
- Add isValidObjectReference() with edge case handling
- Provide foundation for reusable codemod utilities

ShadiBitaraf committed:
refactor(pos-api): modularize and improve transformation logic

- Extract utility functions (getApiAliases, isValidObjectReference)
- Add shouldTransformProperty() helper for cleaner code
- Support complex patterns: this?.api?, getApi()?, function calls
- Improve alias detection with proper destructuring support
- Add comprehensive documentation and comments

ShadiBitaraf committed:
test: add comprehensive test samples for pos-api codemod

- Add shopify-pos-extension.ts with realistic POS patterns
- Add legacy-patterns.ts with class-based and complex patterns
- Add pos-checkout-extension.js with JavaScript patterns
- Add README documentation for testing

ShadiBitaraf committed:
test: add test fixtures for jssg test runner

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+  getMatch(name: string): SgNode | null;
+  replace(text: string): any;
+  findAll(config: { rule: any }): SgNode[];
+}
```
You can import SgNode from `@codemod.com/jssg-types/main`

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+export interface ImportInfo {
+  source: string;
+  specifier: string;
+  node: any;
```
You should refrain from using `any` types. This should probably be a `SgNode<TS, "import_statement">`?

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+  }
+
+  return aliases;
+}
```
This is not taking into account `var` and `let` kinds i think

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+  property: string,
+  method?: string
+): any[] {
+  const usages: any[] = [];
```
no explicit `any`s

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+  });
+  usages.push(...functionCallUsages);
+
+  return usages;
```
This is a good util function to have, but I'd just traverse the member expression like a linked list instead of using string ops. The current code is not readable and hard to follow (and honestly I can't really trust if it covers all the edge cases as it's heavily relying on ast-grep pattern matchers)

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+ * // Find all Polaris imports
+ * const polarisImports = getImports(rootNode, "@shopify/polaris");
+ */
+export function getImports(rootNode: any, packageName: string): ImportInfo[] {
```
no explicit any

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+  packageName: string,
+  importName: string
+): any[] {
+  const nodes: any[] = [];
```
explicit any ❌

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+ * const buttonImports = getNamedImports(rootNode, "@shopify/polaris", "Button");
+ */
+export function getNamedImports(
+  rootNode: any,
```
any

mohebifar commented on utils/ast-utils.ts:
```diff
@@ -0,0 +1,313 @@
+ * @deprecated Use getImports(rootNode, packageName) instead
+ */
+export function getImportSources(
+  rootNode: any,
```
any

ShadiBitaraf committed:
refactor(pos-api-smartgrid): removed extra test files

ShadiBitaraf committed:
chore: updfate project dependencies and minor config updates

ShadiBitaraf committed:
refactor(utils): replace all 'any' types and improve type safety

ShadiBitaraf committed:
refactor(utils): replace string patterns with AST linked-list traversal

- Add getMemberExpressionChain() for robust member expression parsing
- Add matchesMemberPattern() for clean pattern validation
- Replace findMemberExpressions() with proper AST traversal approach

ShadiBitaraf committed:
refactor(utils): take into account 'var' and 'let' patterns too

ShadiBitaraf committed:
refactor: add robust import checking system and eliminate code duplication

- Add generic hasImportsMatching() function with flexible pattern matching (includes/startsWith/exact)
- Add isPOSUIExtensionsFile() for comprehensive POS UI Extensions detection (old and new packages)
- Remove duplicate utility functions from pos-api-smartgrid codemod
- Import reusable functions from utils/ast-utils.ts instead of local duplicates

ShadiBitaraf committed:
chore: update pnpm lockfile
